apps/lockfile-explorer-web/src/containers/LockfileEntryDetailsView/index.tsx

```diff
@@ -231,6 +258,24 @@ export const LockfileEntryDetailsView = (): JSX.Element | ReactNull => {
     );
   };
 
+  const getDependencyInfo = (
+    rawEntryId: string,
+    entryPackageName: string
+  ): { type: DependencyKey; version: string } | undefined => {
+    const packageJson = directRefsPackageJSON.get(rawEntryId);
+    if (!packageJson) return undefined;
+
+    const dependencyTypes = [DependencyKey.DEP, DependencyKey.DEVDEP, DependencyKey.PEERDEP];
+
+    for (const type of dependencyTypes) {
```

Comment:
If I understand correctly, this PR is displaying the original SemVer range (e.g. `"^1.2.3 || ^2.3.4"`) from package.json before PNPM chose a specific version (e.g. `2.5.6`).  

For workspace projects, this information seems to be already present in the PNPM lockfile under the `specifiers` field. Example from Lockfile Explorer:

```yaml
  ../../apps/lockfile-explorer:
    specifiers:
      '@microsoft/rush-lib': workspace:*
      '@rushstack/eslint-config': workspace:*
      '@rushstack/heft': workspace:*
      '@rushstack/heft-node-rig': workspace:*
      '@rushstack/lockfile-explorer-web': workspace:*
      '@rushstack/node-core-library': workspace:*
      '@types/cors': ~2.8.12
      '@types/express': 4.17.13
      '@types/heft-jest': 1.0.1
      '@types/js-yaml': 3.12.1
      '@types/node': 14.18.36
      '@types/update-notifier': ~6.0.1
      colors: ~1.2.1
      cors: ~2.8.5
      express: 4.18.1
      js-yaml: ~3.13.1
      open: ~8.4.0
      update-notifier: ~5.1.0
    dependencies:
      '@rushstack/node-core-library': link:../../libraries/node-core-library
      '@types/express': 4.17.13
      colors: 1.2.5
      cors: 2.8.5
      express: 4.18.1
      js-yaml: 3.13.1
      open: 8.4.0
      update-notifier: 5.1.0
    devDependencies:
      '@microsoft/rush-lib': link:../../libraries/rush-lib
      '@rushstack/eslint-config': link:../../eslint/eslint-config
      '@rushstack/heft': link:../heft
      '@rushstack/heft-node-rig': link:../../rigs/heft-node-rig
      '@rushstack/lockfile-explorer-web': link:../lockfile-explorer-web
      '@types/cors': 2.8.13
      '@types/heft-jest': 1.0.1
      '@types/js-yaml': 3.12.1
      '@types/node': 14.18.36
      '@types/update-notifier': 6.0.1
```
But the lockfile doesn't have a `specifiers` section for external NPM packages.  I think the reason is that external NPM packages are considered to be immutable, whereas a workspace package.json may have been manually modified since the last time that `rush update` was run.

For workspace projects, should we use the value from the lockfile instead of reading package.json?  🤔  I think maybe it is more intuitive to always read package.json, as you are already doing in this PR. I don't have a strong opinion about it, though.

Reply:
I don't know if it's necessary to split it into two parts for reading, such as declaring workspace and then matching the corresponding version in reverse. However, this will indeed reduce the performance waste of file reading. I want to handle it in a unified way.